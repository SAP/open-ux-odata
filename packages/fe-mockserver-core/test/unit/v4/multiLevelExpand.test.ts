import { readFileSync } from 'fs';
import { join } from 'path';
import { DataAccess } from '../../../src/data/dataAccess';
import { ODataMetadata } from '../../../src/data/metadata';
import FileSystemLoader from '../../../src/plugins/fileSystemLoader';
import CDSMetadataProvider from '@sap-ux/fe-mockserver-plugin-cds';
import type { ServiceConfig } from '../../../src';
import ODataRequest from '../../../src/request/odataRequest';

jest.setTimeout(3600000);
describe('Data Access with $expand spanning multiple levels', () => {
    let dataAccess!: DataAccess;
    let metadata!: ODataMetadata;
    const baseUrl = '/MultiLevelExpand';
    const fileLoader = new FileSystemLoader();
    const metadataProvider = new CDSMetadataProvider(fileLoader);
    beforeAll(async () => {
        const baseDir = join(__dirname, 'services', 'multiLevelExpand');

        const cdsContent = readFileSync(join(baseDir, 'service.cds'), 'utf8');
        const edmx = await metadataProvider.loadMetadata(join(baseDir, 'service.cds'));

        metadata = await ODataMetadata.parse(edmx, baseUrl + '/$metadata');
        dataAccess = new DataAccess({ mockdataPath: baseDir } as ServiceConfig, metadata, fileLoader);
    });
    test('List: root level only', async () => {
        const odataRequest = new ODataRequest({ method: 'GET', url: '/A' }, dataAccess);
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: root level only (with $select)', async () => {
        const odataRequest = new ODataRequest({ method: 'GET', url: '/A?$select=ID,value' }, dataAccess);
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 1 level, 1:1', async () => {
        const odataRequest = new ODataRequest({ method: 'GET', url: '/A?$expand=_toOne' }, dataAccess);
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 1 level, 1:n', async () => {
        const odataRequest = new ODataRequest({ method: 'GET', url: '/A?$expand=_toMany' }, dataAccess);
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:1 --> 1:1', async () => {
        const odataRequest = new ODataRequest({ method: 'GET', url: '/A?$expand=_toOne($expand=_toOne)' }, dataAccess);
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:n --> 1:1', async () => {
        const odataRequest = new ODataRequest({ method: 'GET', url: '/A?$expand=_toMany($expand=_toOne)' }, dataAccess);
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:1 --> 1:n', async () => {
        const odataRequest = new ODataRequest({ method: 'GET', url: '/A?$expand=_toOne($expand=_toMany)' }, dataAccess);
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:n --> 1:n', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toMany($expand=_toMany)' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:1 (with $select) --> 1:1', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toOne($expand=_toOne;$select=value)' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:n (with $select) --> 1:1', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toMany($expand=_toOne;$select=value)' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:1 (with $select) --> 1:n', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toOne($expand=_toMany;$select=value)' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:n (with $select) --> 1:n', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toMany($expand=_toMany;$select=value)' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:1 --> 1:1 (with $select)', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toOne($expand=_toOne($select=value))' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:n --> 1:1 (with $select)', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toMany($expand=_toOne($select=value))' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:1 --> 1:n (with $select)', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toOne($expand=_toMany($select=value))' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:n --> 1:n (with $select)', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toMany($expand=_toMany($select=value))' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:1 (with $select) --> 1:1 (with $select)', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toOne($expand=_toOne($select=value);$select=value)' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:n (with $select) --> 1:1 (with $select)', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toMany($expand=_toOne($select=value);$select=value)' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:1 (with $select) --> 1:n (with $select)', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toOne($expand=_toMany($select=value);$select=value)' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
    test('List: 2 levels, 1:n (with $select) --> 1:n (with $select)', async () => {
        const odataRequest = new ODataRequest(
            { method: 'GET', url: '/A?$expand=_toMany($expand=_toMany($select=value);$select=value)' },
            dataAccess
        );
        const data = await dataAccess.getData(odataRequest);
        expect(data).toMatchSnapshot();
    });
});
