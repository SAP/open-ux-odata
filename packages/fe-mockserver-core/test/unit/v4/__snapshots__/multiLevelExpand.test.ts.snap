// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Access with $expand spanning multiple levels List: 1 level, 1:1 1`] = `
[
  {
    "ID": "A-0",
    "_toOne": {
      "ID": "B-0",
      "_back_ID": "A-0",
      "_toOne_ID": "C-0",
      "value": "B",
    },
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne": {
      "ID": "B-1",
      "value": "B1",
    },
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne": {
      "ID": "B-2-1",
      "_toOne_ID": "C-0",
      "value": "B21",
    },
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne": {
      "ID": "B-3",
      "_back_ID": "A-3",
      "_toOne_ID": "",
      "value": "B3",
    },
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 1 level, 1:n 1`] = `
[
  {
    "ID": "A-0",
    "_toMany": [
      {
        "ID": "B-0",
        "_back_ID": "A-0",
        "_toOne_ID": "C-0",
        "value": "B",
      },
    ],
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toMany": [
      {
        "ID": "B-1",
        "value": "B1",
      },
      {
        "ID": "B-2",
        "value": "B2",
      },
    ],
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toMany": [
      {
        "ID": "B-2-1",
        "_toOne_ID": "C-0",
        "value": "B21",
      },
      {
        "ID": "B-2-2",
        "_toOne_ID": "C-0",
        "value": "B22",
      },
    ],
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toMany": [
      {
        "ID": "B-3",
        "_back_ID": "A-3",
        "_toOne_ID": "",
        "value": "B3",
      },
    ],
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:1 (with $select) --> 1:1 (with $select) 1`] = `
[
  {
    "ID": "A-0",
    "_toOne": {
      "ID": "B-0",
      "_toOne": {
        "ID": "C-0",
        "value": "C",
      },
      "value": "B",
    },
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne": {
      "ID": "B-1",
      "_toOne": {
        "ID": "C-1",
        "value": "C1",
      },
      "value": "B1",
    },
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne": {
      "ID": "B-2-1",
      "_toOne": {
        "ID": "C-0",
        "value": "C",
      },
      "value": "B21",
    },
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne": {
      "ID": "B-3",
      "_toOne": {
        "ID": "C-2",
        "value": "C2",
      },
      "value": "B3",
    },
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:1 (with $select) --> 1:1 1`] = `
[
  {
    "ID": "A-0",
    "_toOne": {
      "ID": "B-0",
      "_toOne": {
        "ID": "C-0",
        "_back_ID": "B-0",
        "value": "C",
      },
      "value": "B",
    },
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne": {
      "ID": "B-1",
      "_toOne": {
        "ID": "C-1",
        "value": "C1",
      },
      "value": "B1",
    },
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne": {
      "ID": "B-2-1",
      "_toOne": {
        "ID": "C-0",
        "_back_ID": "B-0",
        "value": "C",
      },
      "value": "B21",
    },
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne": {
      "ID": "B-3",
      "_toOne": {
        "ID": "C-2",
        "value": "C2",
      },
      "value": "B3",
    },
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:1 (with $select) --> 1:n (with $select) 1`] = `
[
  {
    "ID": "A-0",
    "_toOne": {
      "ID": "B-0",
      "_toMany": [
        {
          "ID": "C-0",
          "value": "C",
        },
      ],
      "value": "B",
    },
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne": {
      "ID": "B-1",
      "_toMany": [],
      "value": "B1",
    },
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne": {
      "ID": "B-2-1",
      "_toMany": [
        {
          "ID": "C-2-1",
          "value": "C21",
        },
        {
          "ID": "C-2-2",
          "value": "C22",
        },
      ],
      "value": "B21",
    },
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne": {
      "ID": "B-3",
      "_toMany": [
        {
          "ID": "C-2",
          "value": "C2",
        },
        {
          "ID": "C-3",
          "value": "C3",
        },
      ],
      "value": "B3",
    },
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:1 (with $select) --> 1:n 1`] = `
[
  {
    "ID": "A-0",
    "_toOne": {
      "ID": "B-0",
      "_toMany": [
        {
          "ID": "C-0",
          "_back_ID": "B-0",
          "value": "C",
        },
      ],
      "value": "B",
    },
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne": {
      "ID": "B-1",
      "_toMany": [],
      "value": "B1",
    },
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne": {
      "ID": "B-2-1",
      "_toMany": [
        {
          "ID": "C-2-1",
          "value": "C21",
        },
        {
          "ID": "C-2-2",
          "value": "C22",
        },
      ],
      "value": "B21",
    },
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne": {
      "ID": "B-3",
      "_toMany": [
        {
          "ID": "C-2",
          "value": "C2",
        },
        {
          "ID": "C-3",
          "value": "C3",
        },
      ],
      "value": "B3",
    },
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:1 --> 1:1 (with $select) 1`] = `
[
  {
    "ID": "A-0",
    "_toOne": {
      "ID": "B-0",
      "_back_ID": "A-0",
      "_toOne": {
        "ID": "C-0",
        "value": "C",
      },
      "_toOne_ID": "C-0",
      "value": "B",
    },
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne": {
      "ID": "B-1",
      "_toOne": {
        "ID": "C-1",
        "value": "C1",
      },
      "value": "B1",
    },
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne": {
      "ID": "B-2-1",
      "_toOne": {
        "ID": "C-0",
        "value": "C",
      },
      "_toOne_ID": "C-0",
      "value": "B21",
    },
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne": {
      "ID": "B-3",
      "_back_ID": "A-3",
      "_toOne": {
        "ID": "C-2",
        "value": "C2",
      },
      "_toOne_ID": "",
      "value": "B3",
    },
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:1 --> 1:1 1`] = `
[
  {
    "ID": "A-0",
    "_toOne": {
      "ID": "B-0",
      "_back_ID": "A-0",
      "_toOne": {
        "ID": "C-0",
        "_back_ID": "B-0",
        "value": "C",
      },
      "_toOne_ID": "C-0",
      "value": "B",
    },
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne": {
      "ID": "B-1",
      "_toOne": {
        "ID": "C-1",
        "value": "C1",
      },
      "value": "B1",
    },
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne": {
      "ID": "B-2-1",
      "_toOne": {
        "ID": "C-0",
        "_back_ID": "B-0",
        "value": "C",
      },
      "_toOne_ID": "C-0",
      "value": "B21",
    },
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne": {
      "ID": "B-3",
      "_back_ID": "A-3",
      "_toOne": {
        "ID": "C-2",
        "value": "C2",
      },
      "_toOne_ID": "",
      "value": "B3",
    },
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:1 --> 1:n (with $select) 1`] = `
[
  {
    "ID": "A-0",
    "_toOne": {
      "ID": "B-0",
      "_back_ID": "A-0",
      "_toMany": [
        {
          "ID": "C-0",
          "value": "C",
        },
      ],
      "_toOne_ID": "C-0",
      "value": "B",
    },
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne": {
      "ID": "B-1",
      "_toMany": [],
      "value": "B1",
    },
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne": {
      "ID": "B-2-1",
      "_toMany": [
        {
          "ID": "C-2-1",
          "value": "C21",
        },
        {
          "ID": "C-2-2",
          "value": "C22",
        },
      ],
      "_toOne_ID": "C-0",
      "value": "B21",
    },
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne": {
      "ID": "B-3",
      "_back_ID": "A-3",
      "_toMany": [
        {
          "ID": "C-2",
          "value": "C2",
        },
        {
          "ID": "C-3",
          "value": "C3",
        },
      ],
      "_toOne_ID": "",
      "value": "B3",
    },
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:1 --> 1:n 1`] = `
[
  {
    "ID": "A-0",
    "_toOne": {
      "ID": "B-0",
      "_back_ID": "A-0",
      "_toMany": [
        {
          "ID": "C-0",
          "_back_ID": "B-0",
          "value": "C",
        },
      ],
      "_toOne_ID": "C-0",
      "value": "B",
    },
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne": {
      "ID": "B-1",
      "_toMany": [],
      "value": "B1",
    },
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne": {
      "ID": "B-2-1",
      "_toMany": [
        {
          "ID": "C-2-1",
          "value": "C21",
        },
        {
          "ID": "C-2-2",
          "value": "C22",
        },
      ],
      "_toOne_ID": "C-0",
      "value": "B21",
    },
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne": {
      "ID": "B-3",
      "_back_ID": "A-3",
      "_toMany": [
        {
          "ID": "C-2",
          "value": "C2",
        },
        {
          "ID": "C-3",
          "value": "C3",
        },
      ],
      "_toOne_ID": "",
      "value": "B3",
    },
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:n (with $select) --> 1:1 (with $select) 1`] = `
[
  {
    "ID": "A-0",
    "_toMany": [
      {
        "ID": "B-0",
        "_toOne": {
          "ID": "C-0",
          "value": "C",
        },
        "value": "B",
      },
    ],
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toMany": [
      {
        "ID": "B-1",
        "_toOne": {
          "ID": "C-1",
          "value": "C1",
        },
        "value": "B1",
      },
      {
        "ID": "B-2",
        "_toOne": {
          "ID": "C-1",
          "value": "C1",
        },
        "value": "B2",
      },
    ],
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toMany": [
      {
        "ID": "B-2-1",
        "_toOne": {
          "ID": "C-0",
          "value": "C",
        },
        "value": "B21",
      },
      {
        "ID": "B-2-2",
        "_toOne": {
          "ID": "C-0",
          "value": "C",
        },
        "value": "B22",
      },
    ],
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toMany": [
      {
        "ID": "B-3",
        "_toOne": {
          "ID": "C-2",
          "value": "C2",
        },
        "value": "B3",
      },
    ],
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:n (with $select) --> 1:1 1`] = `
[
  {
    "ID": "A-0",
    "_toMany": [
      {
        "ID": "B-0",
        "_toOne": {
          "ID": "C-0",
          "_back_ID": "B-0",
          "value": "C",
        },
        "value": "B",
      },
    ],
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toMany": [
      {
        "ID": "B-1",
        "_toOne": {
          "ID": "C-1",
          "value": "C1",
        },
        "value": "B1",
      },
      {
        "ID": "B-2",
        "_toOne": {
          "ID": "C-1",
          "value": "C1",
        },
        "value": "B2",
      },
    ],
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toMany": [
      {
        "ID": "B-2-1",
        "_toOne": {
          "ID": "C-0",
          "_back_ID": "B-0",
          "value": "C",
        },
        "value": "B21",
      },
      {
        "ID": "B-2-2",
        "_toOne": {
          "ID": "C-0",
          "_back_ID": "B-0",
          "value": "C",
        },
        "value": "B22",
      },
    ],
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toMany": [
      {
        "ID": "B-3",
        "_toOne": {
          "ID": "C-2",
          "value": "C2",
        },
        "value": "B3",
      },
    ],
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:n (with $select) --> 1:n (with $select) 1`] = `
[
  {
    "ID": "A-0",
    "_toMany": [
      {
        "ID": "B-0",
        "_toMany": [
          {
            "ID": "C-0",
            "value": "C",
          },
        ],
        "value": "B",
      },
    ],
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toMany": [
      {
        "ID": "B-1",
        "_toMany": [
          {
            "ID": "C-1-1",
            "value": "C11",
          },
          {
            "ID": "C-1-2",
            "value": "C12",
          },
        ],
        "value": "B1",
      },
      {
        "ID": "B-2",
        "_toMany": [],
        "value": "B2",
      },
    ],
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toMany": [
      {
        "ID": "B-2-1",
        "_toMany": [],
        "value": "B21",
      },
      {
        "ID": "B-2-2",
        "_toMany": [],
        "value": "B22",
      },
    ],
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toMany": [
      {
        "ID": "B-3",
        "_toMany": [
          {
            "ID": "C-2",
            "value": "C2",
          },
          {
            "ID": "C-3",
            "value": "C3",
          },
        ],
        "value": "B3",
      },
    ],
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:n (with $select) --> 1:n 1`] = `
[
  {
    "ID": "A-0",
    "_toMany": [
      {
        "ID": "B-0",
        "_toMany": [
          {
            "ID": "C-0",
            "_back_ID": "B-0",
            "value": "C",
          },
        ],
        "value": "B",
      },
    ],
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toMany": [
      {
        "ID": "B-1",
        "_toMany": [
          {
            "ID": "C-1-1",
            "value": "C11",
          },
          {
            "ID": "C-1-2",
            "value": "C12",
          },
        ],
        "value": "B1",
      },
      {
        "ID": "B-2",
        "_toMany": [],
        "value": "B2",
      },
    ],
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toMany": [
      {
        "ID": "B-2-1",
        "_toMany": [],
        "value": "B21",
      },
      {
        "ID": "B-2-2",
        "_toMany": [],
        "value": "B22",
      },
    ],
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toMany": [
      {
        "ID": "B-3",
        "_toMany": [
          {
            "ID": "C-2",
            "value": "C2",
          },
          {
            "ID": "C-3",
            "value": "C3",
          },
        ],
        "value": "B3",
      },
    ],
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:n --> 1:1 (with $select) 1`] = `
[
  {
    "ID": "A-0",
    "_toMany": [
      {
        "ID": "B-0",
        "_back_ID": "A-0",
        "_toOne": {
          "ID": "C-0",
          "value": "C",
        },
        "_toOne_ID": "C-0",
        "value": "B",
      },
    ],
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toMany": [
      {
        "ID": "B-1",
        "_toOne": {
          "ID": "C-1",
          "value": "C1",
        },
        "value": "B1",
      },
      {
        "ID": "B-2",
        "_toOne": {
          "ID": "C-1",
          "value": "C1",
        },
        "value": "B2",
      },
    ],
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toMany": [
      {
        "ID": "B-2-1",
        "_toOne": {
          "ID": "C-0",
          "value": "C",
        },
        "_toOne_ID": "C-0",
        "value": "B21",
      },
      {
        "ID": "B-2-2",
        "_toOne": {
          "ID": "C-0",
          "value": "C",
        },
        "_toOne_ID": "C-0",
        "value": "B22",
      },
    ],
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toMany": [
      {
        "ID": "B-3",
        "_back_ID": "A-3",
        "_toOne": {
          "ID": "C-2",
          "value": "C2",
        },
        "_toOne_ID": "",
        "value": "B3",
      },
    ],
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:n --> 1:1 1`] = `
[
  {
    "ID": "A-0",
    "_toMany": [
      {
        "ID": "B-0",
        "_back_ID": "A-0",
        "_toOne": {
          "ID": "C-0",
          "_back_ID": "B-0",
          "value": "C",
        },
        "_toOne_ID": "C-0",
        "value": "B",
      },
    ],
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toMany": [
      {
        "ID": "B-1",
        "_toOne": {
          "ID": "C-1",
          "value": "C1",
        },
        "value": "B1",
      },
      {
        "ID": "B-2",
        "_toOne": {
          "ID": "C-1",
          "value": "C1",
        },
        "value": "B2",
      },
    ],
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toMany": [
      {
        "ID": "B-2-1",
        "_toOne": {
          "ID": "C-0",
          "_back_ID": "B-0",
          "value": "C",
        },
        "_toOne_ID": "C-0",
        "value": "B21",
      },
      {
        "ID": "B-2-2",
        "_toOne": {
          "ID": "C-0",
          "_back_ID": "B-0",
          "value": "C",
        },
        "_toOne_ID": "C-0",
        "value": "B22",
      },
    ],
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toMany": [
      {
        "ID": "B-3",
        "_back_ID": "A-3",
        "_toOne": {
          "ID": "C-2",
          "value": "C2",
        },
        "_toOne_ID": "",
        "value": "B3",
      },
    ],
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:n --> 1:n (with $select) 1`] = `
[
  {
    "ID": "A-0",
    "_toMany": [
      {
        "ID": "B-0",
        "_back_ID": "A-0",
        "_toMany": [
          {
            "ID": "C-0",
            "value": "C",
          },
        ],
        "_toOne_ID": "C-0",
        "value": "B",
      },
    ],
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toMany": [
      {
        "ID": "B-1",
        "_toMany": [
          {
            "ID": "C-1-1",
            "value": "C11",
          },
          {
            "ID": "C-1-2",
            "value": "C12",
          },
        ],
        "value": "B1",
      },
      {
        "ID": "B-2",
        "_toMany": [],
        "value": "B2",
      },
    ],
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toMany": [
      {
        "ID": "B-2-1",
        "_toMany": [],
        "_toOne_ID": "C-0",
        "value": "B21",
      },
      {
        "ID": "B-2-2",
        "_toMany": [],
        "_toOne_ID": "C-0",
        "value": "B22",
      },
    ],
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toMany": [
      {
        "ID": "B-3",
        "_back_ID": "A-3",
        "_toMany": [
          {
            "ID": "C-2",
            "value": "C2",
          },
          {
            "ID": "C-3",
            "value": "C3",
          },
        ],
        "_toOne_ID": "",
        "value": "B3",
      },
    ],
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: 2 levels, 1:n --> 1:n 1`] = `
[
  {
    "ID": "A-0",
    "_toMany": [
      {
        "ID": "B-0",
        "_back_ID": "A-0",
        "_toMany": [
          {
            "ID": "C-0",
            "_back_ID": "B-0",
            "value": "C",
          },
        ],
        "_toOne_ID": "C-0",
        "value": "B",
      },
    ],
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toMany": [
      {
        "ID": "B-1",
        "_toMany": [
          {
            "ID": "C-1-1",
            "value": "C11",
          },
          {
            "ID": "C-1-2",
            "value": "C12",
          },
        ],
        "value": "B1",
      },
      {
        "ID": "B-2",
        "_toMany": [],
        "value": "B2",
      },
    ],
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toMany": [
      {
        "ID": "B-2-1",
        "_toMany": [],
        "_toOne_ID": "C-0",
        "value": "B21",
      },
      {
        "ID": "B-2-2",
        "_toMany": [],
        "_toOne_ID": "C-0",
        "value": "B22",
      },
    ],
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toMany": [
      {
        "ID": "B-3",
        "_back_ID": "A-3",
        "_toMany": [
          {
            "ID": "C-2",
            "value": "C2",
          },
          {
            "ID": "C-3",
            "value": "C3",
          },
        ],
        "_toOne_ID": "",
        "value": "B3",
      },
    ],
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: root level only (with $select) 1`] = `
[
  {
    "ID": "A-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "value": "A3",
  },
]
`;

exports[`Data Access with $expand spanning multiple levels List: root level only 1`] = `
[
  {
    "ID": "A-0",
    "_toOne_ID": "B-0",
    "value": "A",
  },
  {
    "ID": "A-1",
    "_toOne_ID": "",
    "value": "A1",
  },
  {
    "ID": "A-2",
    "_toOne_ID": "",
    "value": "A2",
  },
  {
    "ID": "A-3",
    "_toOne_ID": "B-3",
    "value": "A3",
  },
]
`;
